src/calcium/qqbar.jl

```diff
@@ -0,0 +1,382 @@
+^(a::qqbar, b::Int) = a ^ qqbar(b)
+
+###############################################################################
+#
+#   Exact division
+#
+###############################################################################
+
+function divexact(a::qqbar, b::qqbar)
+   iszero(b) && throw(DivideError())
+   z = qqbar()
+   ccall((:qqbar_div, libcalcium), Nothing,
+         (Ref{qqbar}, Ref{qqbar}, Ref{qqbar}), z, a, b)
+   return z
+end
+
+div(a::qqbar, b::qqbar) = divexact(a, b)
+
+###############################################################################
+#
+#   Comparison
+#
+###############################################################################
+
+function ==(a::qqbar, b::qqbar)
+   return Bool(ccall((:qqbar_equal, libcalcium), Cint,
+                (Ref{qqbar}, Ref{qqbar}), a, b))
+end
+
+###############################################################################
+#
+#   Roots
+#
+###############################################################################
+
+function sqrt(a::qqbar)
+   z = qqbar()
+   ccall((:qqbar_sqrt, libcalcium), Nothing, (Ref{qqbar}, Ref{qqbar}), z, a)
+   return z
+end
+
+function root(a::qqbar, n::Int)
+   n <= 0 && throw(DomainError(n))
+   z = qqbar()
+   ccall((:qqbar_root_ui, libcalcium), Nothing, (Ref{qqbar}, Ref{qqbar}, UInt), z, a, n)
+   return z
+end
+
+function root_of_unity(C::CalciumQQBarField, n::Int)
+   n <= 0 && throw(DomainError(n))
+   z = qqbar()
+   ccall((:qqbar_root_of_unity, libcalcium), Nothing, (Ref{qqbar}, Int, UInt), z, 1, n)
+   return z
+end
+
+function root_of_unity(C::CalciumQQBarField, n::Int, k::Int)
+   n <= 0 && throw(DomainError(n))
+   z = qqbar()
+   ccall((:qqbar_root_of_unity, libcalcium), Nothing, (Ref{qqbar}, Int, UInt), z, k, n)
+   return z
+end
+
+function qqbar_vec(n::Int)
+   return ccall((:_qqbar_vec_init, libcalcium), Ptr{qqbar_struct}, (Int,), n)
+end
+
+function array(R::CalciumQQBarField, v::Ptr{qqbar_struct}, n::Int)
+   r = Vector{qqbar}(undef, n)
+   for i=1:n
+       r[i] = R()
+       ccall((:qqbar_set, libcalcium), Nothing, (Ref{qqbar}, Ptr{qqbar_struct}),
+           r[i], v + (i-1)*sizeof(qqbar_struct))
+   end
+   return r
+end
+
+function qqbar_vec_clear(v::Ptr{qqbar_struct}, n::Int)
+   ccall((:_qqbar_vec_clear, libcalcium), Nothing, (Ptr{qqbar_struct}, Int), v, n)
+end
+
+function roots(f::fmpz_poly, R::CalciumQQBarField)
+    deg = degree(f)
+
+    if deg <= 0
+        return Array{qqbar}(undef, 0)
+    end
+
+    roots = qqbar_vec(deg)
+    ccall((:qqbar_roots_fmpz_poly, libcalcium), Nothing, (Ptr{qqbar_struct}, Ref{fmpz_poly}, Int), roots, f, 0)
+
+    res = array(R, roots, deg)
+    qqbar_vec_clear(roots, deg)
+    return res
+end
+
+# todo: move this
+function numerator(a::fmpq_poly)
+   z = fmpz_poly()
+   ccall((:fmpq_poly_get_numerator, libflint), Nothing,
+         (Ref{fmpz_poly}, Ref{fmpq_poly}), z, a)
+   return z
+end
+
+function roots(f::fmpq_poly, R::CalciumQQBarField)
+    return roots(numerator(f), R)
+end
```

Comment:
```suggestion
function _numerator(a::fmpq_poly)
   z = fmpz_poly()
   ccall((:fmpq_poly_get_numerator, libflint), Nothing,
         (Ref{fmpz_poly}, Ref{fmpq_poly}), z, a)
   return z
end

function roots(f::fmpq_poly, R::CalciumQQBarField)
    return roots(_numerator(f), R)
end
```
This function should not be user facing.

Reply:
Actually, I thought I should move it to fmpq_poly.jl and document it! We already have denominator() for fmpq_poly.

Reply:
This is an illegal polynomial because it has no parents. Users should not be able to create it easily.